src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs

```diff
@@ -41,6 +49,145 @@ private WindowsIdentity()
             : base(null, null, null, ClaimTypes.Name, ClaimTypes.GroupSid)
         { }
 
+        /// <summary>
+        /// Initializes a new instance of the WindowsIdentity class for the user represented by the specified User Principal Name (UPN).
+        /// </summary>
+        /// <remarks>
+        /// Unlike the desktop version, we connect to Lsa only as an untrusted caller. We do not attempt to explot Tcb privilege or adjust the current
+        /// thread privilege to include Tcb.
+        /// </remarks>
+        public WindowsIdentity(string sUserPrincipalName)
+            : base(null, null, null, ClaimTypes.Name, ClaimTypes.GroupSid)
+        {
+            // Desktop compat: See comments below for why we don't validate sUserPrincipalName.
+
+            using (SafeLsaHandle lsaHandle = ConnectToLsa())
+            {
+                int packageId = LookupAuthenticationPackage(lsaHandle, Interop.SspiCli.AuthenticationPackageNames.MICROSOFT_KERBEROS_NAME_A);
+
+                // 8 byte or less name that indicates the source of the access token. This choice of name is visible to callers through the native GetTokenInformation() api
+                // so we'll use the same name the CLR used even though we're not actually the "CLR."
+                byte[] sourceName = { (byte)'C', (byte)'L', (byte)'R', (byte)0 };
+
+                TOKEN_SOURCE sourceContext;
+                if (!Interop.SecurityBase.AllocateLocallyUniqueId(out sourceContext.SourceIdentifier))
+                    throw new SecurityException(Interop.mincore.GetMessage(Marshal.GetLastWin32Error()));
+                sourceContext.SourceName = new byte[TOKEN_SOURCE.TOKEN_SOURCE_LENGTH];
+                Array.Copy(sourceName, sourceContext.SourceName, sourceName.Length);
```

Comment:
Nit: it'd be slightly cheaper to do:
```C#
Array.Copy(sourceName, 0, sourceContext.SourceName, 0, sourceName.Length);
```
so that Array.Copy doesn't have to look up the lower bound.